### ISSUE TITLE
Inconsistent behavior in model.parameters() in Pytorch 1.5.0

### ISSUE THREAD

**Issue by scottcantreid** [pytorch/pytorch#37350]
## 🐛 Bug

In previous versions of PyTorch, I was able to iterate through the `model.parameters()` iterator more than once. Now, in PyTorch 1.5.0, I can only iterate a single time through the parameters, then it becomes empty. I'm not sure if this is a bug or a desired feature, but it caused some of my gradient clipping code to break when I made the transition from 1.4.0 to 1.5.0.

## To Reproduce

```python
import torch

def print_par_shapes(parameters, prefix='\t'):
    i = -1
    for i, p in enumerate(pars):
        print('{}Parameter {}:'.format(prefix, i), p.shape)
    if i == -1:
        print('{}Iterator was empty'.format(prefix))

model = torch.nn.Linear(32, 32)
pars = model.parameters()
print('First iteration:')
print_par_shapes(pars)
print('Second iteration:')
print_par_shapes(pars)
```

## Expected behavior

I expect to see the same parameter shapes printed on both iterations through the parameters. In the case of the `Linear(32,32)` model, the expected outputs are:
```
First iteration:
	Parameter 0: torch.Size([32, 32])
	Parameter 1: torch.Size([32])
Second iteration:
	Parameter 0: torch.Size([32, 32])
	Parameter 1: torch.Size([32])
```
However, on the second iteration, the parameters are an empty iterator. I get the following outputs:
```
First iteration:
	Parameter 0: torch.Size([32, 32])
	Parameter 1: torch.Size([32])
Second iteration:
	Iterator was empty
```

## Environment

PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.6
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

**Comment by mrshenli** [comment]
adding "triage-review" tag to check if this change is expected.

**Comment by ezyang** [comment]
I checked the history of parameters on Module and it seems that it has always returned an iterator. So I am not sure how your original code ever worked. But yes, `parameters()` is specified to return an iterator, and iterators are only iterable over once, so this is as expected.
